Add Hi DPI Support for GLFW

ypujante opened the pull request:
This is the PR following the discussion [Support for 4k/retina displays](https://github.com/emscripten-core/emscripten/discussions/20510)

This is also my first PR and I tried to be as conservative as possible, read the contributing notes, ran as many tests I could on my local machine... I have also tried to minimize the amount of changes.

In the end I have implemented support for Hi DPI environment with webgpu/glfw. I can work in adding other frameworks (SDL, glut, etc...) in the future if this PR is successful and you think it would be worthwhile.

As described in the discussion, I ended up adding a property to the `Browser` object called `isHiDPIAware` which is `false` by default and can be set by calling `setHiDPIAware` (which the demos use). It is also set automatically by glfw depending on the value of `Module.isHiDPIAware` (in `glfwInit()`).

I have created 2 public examples demonstrating the changes (which ends up using webgpu/glfw under the cover).

* [Resizable (no shell)](https://pongasoft.github.io/imgui_raylib_cmake/resizable/imgui_raylib_cmake.html)
* [Non-resizable (with shell)](https://pongasoft.github.io/imgui_raylib_cmake/non-resizable/imgui_raylib_cmake.html)

This demos work with Chrome (119+), Firefox (119+), Safari (16.6+). If you are on a 4k monitor then you will see a difference when checking the `isHiDPIAware` checkbox (it's 100% dynamic). Note that if you move the window from a 4k screen to a 2k screen, the change happens automatically which is pretty cool.

The reason of these 2 demos is because, when there is a shell, the canvas is non-resizable in its own but you can make it fullscreen by clicking the Fullscreen button. The other demo, shows an example where the canvas size is dynamic and take up the full window (completely different code paths, as `onCanvasResize` gets called on every change). In this demo, you can select the Google Chrome menu to go Fullscreen. Note that this use-case was triggering an error in the code `library_glfw.js` line 986 because calling `setSize` with the exact size of the monitor would end up calling `Browser.requestFullscreen` which triggers an error when it is not triggered by the user explicitly (I added a comment in the code with the error).

I did not add any tests because I do not know how to test it since it depends on the tests running on Google Chrome (no other browser to my knowledge supports webgpu) on a 4k monitor. I have no idea if the environment is that way. I will be more than happy to add test(s) if you give me some hints on what to do.

I will also gladly change anything you want me to change to abide by any rules I might have overlooked.

Thank you

sbc100 commented on src/library_browser.js:
```diff
@@ -563,6 +565,12 @@ var LibraryBrowser = {
         var cw = Module["canvas"].width;
         var ch = Module["canvas"].height;
 
+        if(Browser.isHiDPIAware) {
```
Space after `if`

sbc100 commented on src/library_browser.js:
```diff
@@ -620,7 +628,11 @@ var LibraryBrowser = {
     updateResizeListeners() {
       var canvas = Module['canvas'];
-      Browser.resizeListeners.forEach((listener) => listener(canvas.width, canvas.height));
+      if(Browser.isHiDPIAware) {
```
Space after `if`

ypujante replied:
Fixed in latest push

sbc100 commented on src/library_glfw.js:
```diff
@@ -979,14 +983,18 @@ var LibraryGLFW = {
+        //   // Implementation note: callback onCanvasResize will update values accordingly
+        // }
+        Browser.setCanvasSize(width, height);
+        // Implementation note: callback onCanvasResize will update values accordingly
```
This seems like a fix for a separate issue.  Would you make making a separate PR for this?  I wonder if this one might be more testable?

ypujante replied:
Sounds great. I will revert this changes from this PR. I created a separate [issue](https://github.com/emscripten-core/emscripten/issues/20600) for it and will create a separate PR for it.

sbc100 commented on src/library_glfw.js:
```diff
@@ -1160,9 +1169,14 @@ var LibraryGLFW = {
-    Browser.resizeListeners.push((width, height) => {
-       GLFW.onCanvasResize(width, height);
+#if USE_GLFW == 3
+    Browser.isHiDPIAware = !!Module['isHiDPIAware'] || false;
```
If you indent this to be part of the incoming module API then can you add it to `EXTRA_INCOMING_JS_API` in emcc.py and use `{{{ makeModuleReceiveExpr(...) }}}` here?

ypujante replied:
I am not sure if I am doing something wrong, but adding `isHiDPIAware` to `EXTRA_INCOMING_JS_API` then compiling the code generates this:

```javascript
function checkIncomingModuleAPI() {
  ignoredModuleProp('fetchSettings');
  ignoredModuleProp('isHiDPIAware');
}
```

which fails with this message:

```text
Aborted(`Module.isHiDPIAware` was supplied but `isHiDPIAware` not included in INCOMING_MODULE_JS_API)
```

Here is the generated code for `ignoredModuleProp`

```javascript
function ignoredModuleProp(prop) {
  if (Object.getOwnPropertyDescriptor(Module, prop)) {
    abort(`\`Module.${prop}\` was supplied but \`${prop}\` not included in INCOMING_MODULE_JS_API`);
  }
}
```
Any idea?

sbc100 replied:
You would need to this specify `-sINCOMING_MODULE_JS_API=isHiDPIAware` on the command line.

ypujante replied:
Thank you. I am just wondering if it is necessary anymore as it seems a bit complicated for the end user. 

When I started working on the changes that is the idea I came up with because I needed a way to change this flag. Since I added a `setHiDPIAware` on `Browser` which gets injected into `Module` (similarly to `requestFullscreen`, `createContext`, etc...), the calling code can easily make that determination. For example, in the resizable demo I linked to, I have a ImGui checkbox which changes this flag using the API...

So I think I am simply going to remove this line entirely unless you have a different opinion.

sbc100 replied:
sgtm

sbc100 reviewed (commented):
Thanks for details writeup and great looking first PR!

I'm not too familiar with this part of the codebase but the general direction looks good.

ypujante commented:
I just did a push that addresses the issues:

* fixed code style
* reverted changes about fullscreen (created issue #20600)
* reverted `Browser.isHiDPIAware` initialization entirely

sbc100 commented on src/library_browser.js:
```diff
@@ -331,7 +333,7 @@ var LibraryBrowser = {
             Browser.setFullscreenCanvasSize();
           } else {
-            Browser.updateCanvasDimensions(canvas);
+            Browser.setCanvasSize(canvas);
```
This doesn't look right since `setCanvasSize` takes a width and a height, not a canvas object.

ypujante replied:
yes you are right. wth did I do? :(

ypujante replied:
I profusely apologize for this oversight.

I just pushed a fix for it. Ironically the "wrong" code was still doing the "right" thing and that is the reason why I never saw it during my testing:

* calling `Browser.setCanvasSize(canvas)` ends up calling `Browser.updateCanvasDimensions(canvas, canvas);`
* and due to the test `if (wNative && hNative)` (`hNative` is `undefined`), it executes the `if`
* The normal call ends up calling `Browser.updateCanvasDimensions(canvas)` with both `wNative` and `hNative` undefined resulting in the exact same code path being triggered

The joys of javascript...

In any case, the code is now doing the "right" thing while being "right" as well :)

ypujante commented:
@sbc100 Reverting the changes related to issue #20600 makes fullscreen/exit fullscreen no longer work in my test scenarios. Since I reverted the changes I had made, I assume it works like it used to. But since I don't have a good way to verify it, I don't feel very comfortable with the state of the PR at this moment. I will let you know my findings after I spend more time with it.

ypujante commented:
Well good news. I did some more local testing and by using the default shell that comes with emscripten, the "Fullscreen" button works as expected with the changes in this PR. I can also see that all the tests have passed. So I am now confident with the state of this PR.

Of course, the other use case we discussed in issue #20600 still remain a problem, but it is the same issue before this PR. So I am going to spend some time now and try to see if I can address it. But we have decoupled it from this PR which I think was the right call.

sbc100 commented on src/library_browser.js:
```diff
@@ -27,6 +27,7 @@ var LibraryBrowser = {
 #endif
     Module["requestAnimationFrame"] = (func) => Browser.requestAnimationFrame(func);
     Module["setCanvasSize"] = (width, height, noUpdates) => Browser.setCanvasSize(width, height, noUpdates);
+    Module["setHiDPIAware"] = (isHiDPIAware) => Browser.setHiDPIAware(isHiDPIAware);
```
Do we need to expose this on the outer `Module` object like this?

Would it make more sense to have this be a build time setting?  Or is this something one would want to change at runtime?

ypujante replied:
* Since I removed the code that initializes the `Browser.isHiDPIAware` variable there needs to be a way to set it
* I don't believe it can be a build time setting because I do believe that the client code needs to set it based on the environment (and I was definitely planning to have my program offer the user the ability to toggle it if the performance are not satisfactory for example).

sbc100 replied:
Fair enough

sbc100 commented:
@juj WDYT?

ypujante commented:
I can see that github says that there is a conflict that must be resolved. When I click "Resolve conflicts", it doesn't look like I can do anything about it. Am I supposed to fix it on my fork and push again?

sbc100 commented:
Yes, I would fix locally then then push again.  You can use either `merge` or `rebase` as you prefer.

ypujante commented:
@sbc100  I did a merge and pushed it. I am not sure I understand why there are 90 files changed. Let me know if this is normal or if I screwed something up

sbc100 commented:
> @sbc100 I did a merge and pushed it. I am not sure I understand why there are 90 files changed. Let me know if this is normal or if I screwed something up. My [branch](https://github.com/emscripten-core/emscripten/compare/main...ypujante:emscripten:hi-dpi-aware) does not look like that...

I think you screwed something up yes...  Im not sure exactly what but it doesn't look right :-/

ypujante commented:
@sbc100 I believe it is fixed now as the PR only shows the 2 modified files (I used github "Sync" branch on my fork this time). I also checked that the changes to the 2 files are what I was expecting and double checked that the expected behavior is the same on my local test playground.

ypujante commented:
Fixed the merge conflict introduced by the latest check ins and pushed the branch. It looks like this time I did it right the first time...

ypujante commented:
As far as I can tell the failing tests have nothing to do with my changes:

```text
ports:INFO: retrieving port: mpg123 from https://www.mpg123.de/download/mpg123-1.26.2.tar.bz2
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/urllib3/connection.py", line 141, in _new_conn
    (self.host, self.port), self.timeout, **extra_kw)
  File "/usr/lib/python3/dist-packages/urllib3/util/connection.py", line 83, in create_connection
    raise err
  File "/usr/lib/python3/dist-packages/urllib3/util/connection.py", line 73, in create_connection
    sock.connect(sa)
TimeoutError: [Errno 110] Connection timed out

During handling of the above exception, another exception occurred:
```

Do I need to do anything about this?

sbc100 commented:
I restarted the CI

ypujante commented:
After working on issue #20600, I realized that I could write a fairly comprehensive test for this pull request. The idea is to mock the `devicePixelRatio` so that it no longer depends on which screen/hardware the test is being run on.

During the writing of this test, I realized that I needed to add a listener to detect the changes to  `devicePixelRatio` and I follow the [example from the documentation](https://developer.mozilla.org/en-US/docs/Web/API/Window/devicePixelRatio#monitoring_screen_resolution_or_zoom_level_changes) to add such a listener.

I added the test to `test_browser.py` and I assume it gets picked up automatically but please let me know if that is not the case. I will definitely check the results of CI to see if I can see it running in the output...

sbc100 commented on src/library_browser.js:
```diff
@@ -620,7 +630,11 @@ var LibraryBrowser = {
+      if (Browser.isHiDPIAware) {
+        Browser.resizeListeners.forEach((listener) => listener(canvas.clientWidth, canvas.clientHeight, canvas.width, canvas.height));
+      } else {
+        Browser.resizeListeners.forEach((listener) => listener(canvas.width, canvas.height, canvas.width, canvas.height));
```
Can you avoid duplication here by doing something like:

```
var cw =  canvas.width, ch = canvas.height;
if (Browser.isHiDPIAware) {
  cw = canvas.clientWidth;
  hw = canvas.clientHeight;
}
Browser.resizeListeners.forEach((listener) => listener(cw, ch, canvas.width, canvas.height));
```

ypujante replied:
No problem. Changed accordingly

sbc100 commented on src/library_browser.js:
```diff
@@ -678,23 +694,55 @@ var LibraryBrowser = {
+        }
+        if (Browser.isHiDPIAware) {
+          Browser.devicePixelRatioMQS = window.matchMedia("(resolution: " + Browser.getDevicePixelRatio() + "dppx)");
+          Browser.devicePixelRatioMQS.addEventListener("change", Browser.onDevicePixelRatioChange);
```
Use single quotes for strings in this function.   I know we are not consistent everywhere, but we try to use single quotes by default.

ypujante replied:
Done

sbc100 commented on test/test_browser.py:
```diff
@@ -2968,6 +2968,10 @@ def test_glfw_events(self):
 
+  @requires_graphics_hardware
+  def test_glfw3_hi_dpi_aware(self):
+    self.btest('test_glfw3_hi_dpi_aware.c', args=['-sUSE_GLFW=3', '-lglfw', '-lGL'], expected='1')
```
Can you use btest_exit here .. then you don't need any use of `REPORT_RESULT` in your code, you just return 0 from main.

ypujante replied:
I just copy/pasted other glfw tests but easy to change. Done.

sbc100 reviewed (commented):
Nice!

sbc100 commented on test/browser/test_glfw3_hi_dpi_aware.c:
```diff
@@ -0,0 +1,166 @@
+}
+
+int main()
+{
```
Opening brace on same line a function declaration.

ypujante replied:
My bad. Done.

sbc100 commented on test/browser/test_glfw3_hi_dpi_aware.c:
```diff
@@ -0,0 +1,166 @@
+int main()
+{
+  
+    GLFWwindow *window;
```
Is this function 2-space indented?  I can't tell easily from the github UI but it looks like it might be 3?

ypujante replied:
Not sure what happened there. I remember seeing my editor complaining about not being able to process the clang format file. I wonder if that caused this issue.

In any case it is now fixed

sbc100 reviewed (commented):
lgtm % comments

Still waiting on @juj for any input on the addition to the browser API.

ypujante commented:
@sbc100 Looking at the log file of `test-browser-firefox` I am seeing this:

```
  test_glfw3_hi_dpi_aware (test_browser.browser) ... skip (0.000s)
``` 

Does this mean it did not run? And if not why? And what can I do to make it run?

sbc100 commented:
> @sbc100 Looking at the log file of `test-browser-firefox` I am seeing this:
> 
> ```
>   test_glfw3_hi_dpi_aware (test_browser.browser) ... skip (0.000s)
> ```
> 
> Does this mean it did not run? And if not why? And what can I do to make it run?
> 
> Looking at another test run (test-browser-chrome-wasm64) I am seeing this:
> 
> ```
>   test_glfw3_hi_dpi_aware (test_browser.browser64) ... cache:INFO: generating system asset: symbol_lists/28f4d8e7b74de77ad2c287839c894b23f3154bf6.json... (this will be cached in "/root/cache/symbol_lists/28f4d8e7b74de77ad2c287839c894b23f3154bf6.json" for subsequent builds)
> cache:INFO:  - ok
> ok (1.127s)
> ```
> 
> So it looks like it is running at least in some test cases.

The test is marked as `@requires_graphics_hardware` and we don't currently run graphics tests under firefox: https://github.com/emscripten-core/emscripten/blob/898e9bbaf2ed1a322b1476712f0b42443d0af11e/.circleci/config.yml#L368-L370.   It looks like maybe we could try removing that limitation, but for now we have to rely on chrome for graphics tests in CI

ypujante commented:
@sbc100 Thanks for confirming.

The good news is that I can see that it runs fine on chrome in the CI environment (`test-browser-chrome`)

```text
  test_glfw3_hi_dpi_aware (test_browser.browser) ... cache:INFO: generating system asset: symbol_lists/ceafa48ea5ed0f3da70cf72426a36777b536dfe0.json... (this will be cached in "/root/cache/symbol_lists/ceafa48ea5ed0f3da70cf72426a36777b536dfe0.json" for subsequent builds)
cache:INFO:  - ok
ok (0.991s)
```

At least locally on my machine, this test runs fine on Firefox as well. Since Firefox 119 it looks like there is now full support for webgpu from my experience.

ypujante commented:
After making my branch up to date with main, tests are failing which what does not seem to be my issue:

```text
/root/cache/ports/sdl2/SDL-release-2.24.2/src/SDL_assert.c:263:25: error: cast to 'char *' from smaller integer type 'int' [-Werror,-Wint-to-pointer-cast]
  263 |             char *buf = (char *) EM_ASM_INT({
      |                         ^~~~~~~~~~~~~~~~~~~~~
  264 |                 var str =
      |                 ~~~~~~~~~
  265 |                     UTF8ToString($0) + '\n\n' +
      |                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~
  266 |                     'Abort/Retry/Ignore/AlwaysIgnore? [ariA] :';
      |                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  267 |                 var reply = window.prompt(str, "i");
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  268 |                 if (reply === null) {
      |                 ~~~~~~~~~~~~~~~~~~~~~
  269 |                     reply = "i";
      |                     ~~~~~~~~~~~~
  270 |                 }
      |                 ~
  271 |                 return allocate(intArrayFromString(reply), 'i8', ALLOC_NORMAL);
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  272 |             }, message);
      |             ~~~~~~~~~~~
1 error generated.
emcc: error: '/root/emsdk/upstream/bin/clang -target wasm64-unknown-emscripten -fignore-exceptions -fvisibility=default -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr -Werror=implicit-function-declaration --sysroot=/root/cache/sysroot -Xclang -iwithsysroot/include/fakesdl -Xclang -iwithsysroot/include/compat -c -g3 -Werror -O2 -I/root/cache/ports/sdl2/SDL-release-2.24.2 -I/root/cache/sysroot/include/SDL2 /root/cache/ports/sdl2/SDL-release-2.24.2/src/SDL_assert.c -o /root/cache/ports-builds/sdl2/src/SDL_assert.c.o' failed (returned 1)
emcc: error: Subprocess 2/115 failed (returned 1)! (cmdline: /root/project/emcc -c /root/cache/ports/sdl2/SDL-release-2.24.2/src/SDL_assert.c -o /root/cache/ports-builds/sdl2/src/SDL_assert.c.o -g -sSTRICT -Werror -Wno-experimental -sMEMORY64=1 -O2 -I/root/cache/ports/sdl2/SDL-release-2.24.2 -sUSE_SDL=0 -I/root/cache/sysroot/include/SDL2)
None
None
FAIL (0.371s)
  test_sdl2_image_formats (test_browser.browser64) ... skip (0.007s)
  test_sdl2_image_formats_wasmfs (test_browser.browser64) ... skip (0.006s)
  test_sdl2_image_jpeg (test_browser.browser64) ... skip (0.006s)
  test_sdl2_image_memfile (test_browser.browser64) ... cache:INFO: generating port: sysroot/lib/wasm64-emscripten/libSDL2.a... (this will be cached in "/root/cache/sysroot/lib/wasm64-emscripten/libSDL2.a" for subsequent builds)
/root/cache/ports/sdl2/SDL-release-2.24.2/src/SDL_assert.c:263:25: error: cast to 'char *' from smaller integer type 'int' [-Werror,-Wint-to-pointer-cast]
  263 |             char *buf = (char *) EM_ASM_INT({
      |                         ^~~~~~~~~~~~~~~~~~~~~
  264 |                 var str =
      |                 ~~~~~~~~~
  265 |                     UTF8ToString($0) + '\n\n' +
      |                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~
  266 |                     'Abort/Retry/Ignore/AlwaysIgnore? [ariA] :';
      |                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  267 |                 var reply = window.prompt(str, "i");
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  268 |                 if (reply === null) {
      |                 ~~~~~~~~~~~~~~~~~~~~~
  269 |                     reply = "i";
      |                     ~~~~~~~~~~~~
  270 |                 }
      |                 ~
  271 |                 return allocate(intArrayFromString(reply), 'i8', ALLOC_NORMAL);
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  272 |             }, message);
      |             ~~~~~~~~~~~
1 error generated.
emcc: error: '/root/emsdk/upstream/bin/clang -target wasm64-unknown-emscripten -fignore-exceptions -fvisibility=default -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr -Werror=implicit-function-declaration --sysroot=/root/cache/sysroot -Xclang -iwithsysroot/include/fakesdl -Xclang -iwithsysroot/include/compat -c -g3 -Werror -O2 -I/root/cache/ports/sdl2/SDL-release-2.24.2 -I/root/cache/sysroot/include/SDL2 /root/cache/ports/sdl2/SDL-release-2.24.2/src/SDL_assert.c -o /root/cache/ports-builds/sdl2/src/SDL_assert.c.o' failed (returned 1)
emcc: error: Subprocess 2/115 failed (returned 1)! (cmdline: /root/project/emcc -c /root/cache/ports/sdl2/SDL-release-2.24.2/src/SDL_assert.c -o /root/cache/ports-builds/sdl2/src/SDL_assert.c.o -g -sSTRICT -Werror -Wno-experimental -sMEMORY64=1 -O2 -I/root/cache/ports/sdl2/SDL-release-2.24.2 -sUSE_SDL=0 -I/root/cache/sysroot/include/SDL2)
None
None
```

sbc100 commented:
Have you changed `test/test_browser.py`?   That test (`test_sdl2_image`) should not run under browser64 as it is marked with `@no_wasm64('SDL2 + wasm64')`

sbc100 commented on test/test_browser.py:
```diff
@@ -3027,6 +3027,10 @@ def test_glfw_events(self):
     self.btest('test_glfw_events.c', args=['-sUSE_GLFW=3', "-DUSE_GLFW=3", '-lglfw', '-lGL'], expected='1')
 
   @no_wasm64('SDL2 + wasm64')
```
You need to move this decorator back down so it applies to `test_sdl2_image`

ypujante replied:
I guess it is the github automatic merge that screwed it up :(

ypujante commented:
@sbc100 Thank you for pointing out what the merge got wrong. Fixed and tests restarted. I guess the good news is that the tests caught the merge issue...

ypujante commented:
After merging with `main` to pick up latest commits/fix conflicts, the tests are failing:

```text
The install of pkgconfiglite was NOT successful.
Error while running 'C:\ProgramData\chocolatey\lib\pkgconfiglite\tools\chocolateyInstall.ps1'.
 See log for details.
```

kripken commented:
Unfortunately I know nothing about HiDPI myself. If @juj isn't available then one of us would need to take time to learn about these Web APIs I think @sbc100 . If possible it would also be great to find a way to run the `interactive` tests as they might have coverage for window resizing and such, and we don't run them on CI.

ypujante commented:
Let me know if there anything I can do to help with the effort. I understand that my changes have a potential of introducing regression issues, especially for the `Browser` code which is used by multiple libraries (glfw, SDL, glut...). If I have a way myself to run the `interactive` suite of tests, please let me know how I would go about it. It would also be fairly easy, to really make the feature have even less impact by having truly 2 different paths whether `isHiDPIAware` is set or not in all places where I have made modifications. Or make it an "experimental" feature that can be turned on via a compilation flag? 

Talking about potential regression, I was looking at a test yesterday which is run for SDL ([test_sdl_canvas_palette](https://github.com/emscripten-core/emscripten/blob/main/test/browser/test_sdl_canvas_palette.c)): as far as I can tell, this test runs on CI but it does not seem to be testing anything (no `assert` in the code) unless I am missing something. This test passes on my PR but if it doesn't test anything, then it doesn't prove that my changes have not affected SDL for example.

sbc100 commented:
> Let me know if there anything I can do to help with the effort. I understand that my changes have a potential of introducing regression issues, especially for the `Browser` code which is used by multiple libraries (glfw, SDL, glut...). If I have a way myself to run the `interactive` suite of tests, please let me know how I would go about it. It would also be fairly easy, to really make the feature have even less impact by having truly 2 different paths whether `isHiDPIAware` is set or not in all places where I have made modifications. Or make it an "experimental" feature that can be turned on via a compilation flag?
> 
> Talking about potential regression, I was looking at a test yesterday which is run for SDL ([test_sdl_canvas_palette](https://github.com/emscripten-core/emscripten/blob/main/test/browser/test_sdl_canvas_palette.c)): as far as I can tell, this test runs on CI but it does not seem to be testing anything (no `assert` in the code) unless I am missing something. This test passes on my PR but if it doesn't test anything, then it doesn't prove that my changes have not affected SDL for example.

`test_sdl_canvas_palette` is what we call a reftest.  It renders to a canvas and then compares the result to a reference imagine: https://github.com/emscripten-core/emscripten/blob/97f37e125d61bc46fbeb2ad5f54b5d6fdd1b6140/test/test_browser.py#L2263-L2264

ypujante commented:
That's very clever! Thank you.

ypujante commented:
@sbc100 As I was looking into adding Hi DPI support for SDL, I noticed that there were 2 different SDL: the one in emscripten (`library_sdl.js`) and the "ports" directly on the main SDL project. I was not aware of that nor is it clear to me why there are 2.

To my surprise, I also noticed that it seems that the SDL ports already supports Hi DPI as [evidence by this code](https://github.com/libsdl-org/SDL/blob/4722269fb62315640cbdbd8916cfc937ee3603d4/src/video/emscripten/SDL_emscriptenvideo.c#L194). I have not tested that it works, but it's there for sure!

I am not sure there is a huge value in adding HiDPI support to `library_sdl.js` since the port already has support for it.

Since this PR is about adding HiDPI support to GLFW, maybe I can find a way to limit/move the changes I introduced in the `Browser` class  to `library_glfw.js` instead. This way, the potential regression would be far more limited to glfw only rather than the other libraries using `Browser`.

I am thinking closing the PR.

What do you think?

sbc100 commented:
> @sbc100 As I was looking into adding Hi DPI support for SDL, I noticed that there were 2 different SDL: the one in emscripten (`library_sdl.js`) and the "ports" directly on the main SDL project. I was not aware of that nor is it clear to me why there are 2.
> 
> To my surprise, I also noticed that it seems that the SDL ports already supports Hi DPI as [evidence by this code](https://github.com/libsdl-org/SDL/blob/4722269fb62315640cbdbd8916cfc937ee3603d4/src/video/emscripten/SDL_emscriptenvideo.c#L194). I have not tested that it works, but it's there for sure!
> 
> I am not sure there is a huge value in adding HiDPI support to `library_sdl.js` since the port already has support for it.
> 
> Since this PR is about adding HiDPI support to GLFW, maybe I can find a way to limit/move the changes I introduced in the `Browser` class to `library_glfw.js` instead. This way, the potential regression would be far more limited to glfw only rather than the other libraries using `Browser`.
> 
> I am thinking closing the PR.
> 
> What do you think?

Yes there are two versions of SDL.  The one in JavaScripte is SDL1 and you get access to this using `-sUSE_SDL` and the `ports` one is the written in C++ and maintained upstream and is SDL2 and used via `-sUSE_SDL=2`.

I'm not sure what the risks are of regression here... isn't this change supposed to be a no-op for users to don't use the new API?  i.e. I would hope there is no risk of regressions for existing code.

I don't really see a need to close this issue unless you are proposing a new / better way for GLFW users to access HIDPI mode?  

My understanding is that only remaining thing to discuss is whether the new setHiDPIAware() API is the right way to go for controlling this.  Wouldn't that question be the same even if this was purely focused on GLFW?

ypujante commented:
@sbc100 Oh okay, my bad. I was under the impression that what was blocking my PR was the potential for regression since I did end up touching a central class (`Browser`). The way the changes have been written are not 100% no-op (it's close but not 100%).

For example this change:

```javascript
            Browser.updateCanvasDimensions(canvas);
            Browser.updateResizeListeners(); // this was not present before my PR
```

would be executed no matter what. Of course all the tests pass which is reassuring.

If instead, it is just a matter of whether the API change is right or wrong, then yes I will leave it open.

Note that the way SDL2 handles it is via a flag (`SDL_WINDOW_HIGH_PIXEL_DENSITY`). Maybe I could remove the exposed API (in `Module`) entirely and follow a similar approach by using a GLFW flag (for example [`GLFW_SCALE_TO_MONITOR`](https://github.com/emscripten-core/emscripten/blob/c017fc2d6961962ee87ae387462a099242dfbbd2/src/library_glfw.js#L116))?

ypujante commented:
> My understanding is that only remaining thing to discuss is whether the new setHiDPIAware() API is the right way to go for controlling this. Wouldn't that question be the same even if this was purely focused on GLFW?

I believe you are right. My understanding looking at the SDL3 code is that it is done via a (window) flag set when creating the window (`SDL_WINDOW_HIGH_PIXEL_DENSITY`). Not being very familiar with SDL, I do not know if there is a way to change a window flag after it is created but I don't think so.

Instead of adding a `setHiDPIAware` API to Module, we could follow a similar pattern to SDL3 for GLFW and use the `glfwWindowHint` `GLFW_SCALE_TO_MONITOR` instead. The only drawback is that it cannot be changed after the fact, but that would be the same behavior as implemented by SDL3

It is true that the way I implemented it, makes it available to any code right now and so, technically, you can call `Module.setHiDPIAware(true)` for an emscripten application that does not use GLFW (uses glut for example) and that will result in wrong scaling.

juj commented:
Hey, sorry for the delay - I was on a sick leave for a while with a bit nasty flu.

Really cool to see contributions in this area. I think we do want to revise this PR a bit, but to figure what would be the best way, let me first give a bit of background info here:

In the current state, the library_glfw.js, library_glut.js, library_glew.js, library_sdl.js and library_xlib.js, and a few others; and partially **library_browser.js** are castaway/derelict. These libraries were very important in the early days of Emscripten when we were working hard to ensure that it would be easier to get started building code with Emscripten, when we had relatively few codebases successfully compiling to the web. (we were eager to get the first projects working)

Today, we have a lot more code successfully compiling, so the focus has shifted to doing 100% complete ports of libraries, and to paying attention a lot to the code quality that Emscripten generates. So the above libraries hit a bit of a deadend in how they could scale, due to their "ad hoc rewrites" nature, and they went in legacy maintenance mode.

Landing any changes to library_glfw.js, library_glut.js, library_glew.js, library_sdl.js or library_xlib.js, even if experimental, I am not worried about, since there are relatively few users of those libraries, and if people don't use them, they won't receive any code into their projects from those library files. If there are people who want to champion improvements to them, that would be great, and they can have relatively easy time to work on the surface area of these self-contained libraries.

However, library_browser.js is different. Today every non-trivial sized Emscripten compiled project receives a copy of library_browser.js in the build output. And due to how it was originally designed (the `$Browser` object, and the use of `Module`), it does not dead-code eliminate well.

So we want to be very cautious in adding any new code into the `$Browser` or `Module` objects, since a lot of compiled projects out there will then unconditionally emit that code, independent of whether they use the newly added features. 

So initially my recommendation/question here is: could this hi-DPI support be added to GLFW in a manner that did not touch library_browser.js? That way the feature would be self-contained to just GLFW.

(Superficially, the above recommendation might raise an objection on the basis of "well, that wouldn't then be as reusable, would it?", but the thing is, that I don't think that reusing code across GLFW, GLEW, GLUT, SDL is a good idea on the long run. Such reuse creates code dependencies that prevent encapsulation and loses ergonomy in refactoring, and code sharing won't save code bytes in these libraries anyways)

So there is a bit of pressure to keep the `$Browser` object frozen so that it does not increase code size for users - one day we'd like to remove it, but due to its public API nature, it is a bit difficult today.

Recently, work has gone into improving the power of end-user developed library_x.js files as opposed to Emscripten system-provided library_x.js files. (@sbc100  work on __cdeps to enable JS->C dependencies was a big missing feature).

Ultimately the best way to create strong support forward for GLFW would be to do structure that library similar to how I authored support for WebGPU bindings recently at https://github.com/juj/wasm_webgpu: have a standalone repository that implements the needed library and code files: https://github.com/juj/wasm_webgpu/tree/main/lib . This kind of design can be tied with Emscripten's ports system for easy distribution/integration if people would like to develop that. (like SDL 2 for example works today)

So I think before jumping deep into reviewing individual code lines, I'll pose the two questions:
a) would it be possible/make sense to implement this high-dpi support for GLFW inside library_glfw.js standalone so that library_browser.js is not touched, so that code size is not increased for projects that don't use it?
b) how big is your appetite to further develop library_glfw.js? If you're enthusiastic for championing further development, maybe this might be the time to migrate library_glfw.js into its own repository and maybe use the ports system to distribute it? (that's ok if not, just bringing up the idea in general)

ypujante commented:
@juj Sorry to hear you were sick. But happy that you are back :)

First thank you so much for providing feedback/history on the project. This is definitely the kind of details that cannot be extracted by just looking at the source code.

A little bit of my history/background. My most recent project is an ImGui (with raylib backend) desktop (macOs/Windows) [project](https://github.com/pongasoft/re-edit). When I saw that raylib was supporting the web platform via emscripten I was very excited to see if I could run my tool in the browser (which would be quite cool and avoid multiple platforms to support). So I tried it and quickly realized how [low quality rendering](https://github.com/raysan5/raylib/issues/3372) was on this platform.

<img width="348" alt="macOS" src="https://github.com/raysan5/raylib/assets/357116/012cb898-fbda-4254-843b-b3dc1a889e2b">

And so I spent a lot of time trying to figure out how to "fix" this issue which lead to my PR for emscripten to support HiDPI with glfw (raylib uses emscripten/glfw).


> Recently, work has gone into improving the power of end-user developed library_x.js files as opposed to Emscripten system-provided library_x.js files. (@sbc100 work on __cdeps to enable JS->C dependencies was a big missing feature).

This sounds very interesting. Could you point me to some documentation about it?

> So I think before jumping deep into reviewing individual code lines, I'll pose the two questions:
> a) would it be possible/make sense to implement this high-dpi support for GLFW inside library_glfw.js standalone so that library_browser.js is not touched, so that code size is not increased for projects that don't use it?
> b) how big is your appetite to further develop library_glfw.js? If you're enthusiastic for championing further development, maybe this might be the time to migrate library_glfw.js into its own repository and maybe use the ports system to distribute it? (that's ok if not, just bringing up the idea in general)

These are fair and valid questions before you start digging.

For a), based on your description of the project and how "critical" `Browser` is, it would be preferable for me to see how I can implement these changes with no modifications whatsoever to `Browser`. So this is what I am going to be looking into now. So no need to dig into my changes until I report back :)

For b) I am not too sure. That seems like a big effort. Which I wouldn't mind tackling but I am not an expert with glfw in the first place. I can definitely spend some time digging further into the current `library_glfw.js` implementation (I have only looked at the parts that were relevant for this issue and I am not sure what else is there). And I also need to look at your port for webgpu which I suppose would be a good example to follow before I commit to this kind of challenge.

Thank you!

sbc100 commented:
Mostly agree with @juj's feedback.     I would like to confirm/re-iterate that the important thing is not that updating `library_browser.js` is bad, but that growing the `Browser` object (which cannot be DCE'd well) should be avoided.  If you can find a way to make this work without growing the `Browser` object for existing users I think that would be OK.

We should probably create a codesize test that includes `Browser` so we can monitor its size and avoid accidental growth.

sbc100 commented on src/library_glfw.js:
```diff
@@ -1055,6 +1060,13 @@ var LibraryGLFW = {
+      for (var k in GLFW.defaultHints) {
+        GLFW.hints[k] = GLFW.defaultHints[k];
+      }
+    },
```
Looks like this can be rebased to make this change smaller now that https://github.com/emscripten-core/emscripten/pull/20770 has landed?

ypujante replied:
I am in the process of writing details about this changes.. just hold on a secondf on the review

ypujante replied:
code has been rebased and this is gone

ypujante commented:
@juj @sbc100 I just pushed a new version that is 100% self contained within GLFW with 0 changes in `Browser`. I moved all the extra that I had added on `Browser` to `GLFW`. And I "override" exactly 3 calls:

* `requestFullScreen` for the sole reason that I need to also raise an event when `resize` is `false` (if you are ok with this 1 line addition to `Browser` we could move it back to `Browser` which would eliminate this override)
* `updateCanvasDimensions` which is the core of the changes to handle HiDPI
* `calculateMouseEvent` to handle HiDPI

I want your feedback on whether you think this is the "right" direction

Please note that this push is "broken" in the sense that I realized that `Browser.calculateMouseEvent` was modified after I copy/pasted it in `GLFW` so I need to incorporate the changes there. That is the danger/drawback of such a technique...

ypujante commented:
@sbc100 I am done with the changes now and rebased the code to pull the merge you did on the main branch. I updated the feedback in my previous comment. Thank you

sbc100 commented on src/library_glfw.js:
```diff
@@ -88,6 +90,9 @@ var LibraryGLFW = {
   ],
+  $GLFW__postset: `
+    // making setHiDPIAware public for users of glfw (dynamic switching)
+    Module["setHiDPIAware"] = GLFW.setHiDPIAware;`,
```
I think you can just do this:

```
setHiDPIAware: GLFW.setHiDPIAware,
```

Then users who want to export that function can do so like any other library function.   No need to export it by default like this.

ypujante replied:
I don't understand what you are suggesting:

1) if I write `setHiDPIAware: GLFW.setHiDPIAware` isn't it defining a field referring to the same field?
2) how does a user gets access to this function? can you explain what a user would have to do to access it? Right now I write this (in C) `EM_ASM({Module.setHiDPIAware($0);}, isHiDPIAware ? 1 : 0);`. What would I write after the change and is there compilation flags to be provided?

sbc100 replied:
`$setHiDPIAware: GLFW.setHiDPIAware,` would be something you would write at the higher level (sorry I forgot to include the leading dollar sign the first time) not within the `GLFW` object.  It would act like any other JS library function.

You can include JS library functions using `-sDEFAULT_LIBRARY_FUNCS_TO_INCLUDE` on the command line and export them via `-sEXPORTED_RUNTIME_FUNCTIONS` or `-sEXPORTED_FUNCTIONS`.

If you want to call `setHiDPIAware` from an EM_ASM block you wouldn't actually need to the `Module.` prefix at all.   In that case you can just call `setHiDPIAware` and then specify it as a dependency using `EM_JS_DEPS(foo, "$setHiDPIAware");` in your C/C+ source code.

ypujante replied:
> You can include JS library functions using -sDEFAULT_LIBRARY_FUNCS_TO_INCLUDE on the command line and export them via -sEXPORTED_RUNTIME_FUNCTIONS or -sEXPORTED_FUNCTIONS.

Are these options on the client side or something I need to change on the emscripten side?

Is it wise to call it `$setHiDPIAware`? Should it be called `$glfwSetHiDPIAware` or `$glfwEmscriptenSetHiDPIAware` so that there are less risks of collision?

sbc100 replied:
> > > You can include JS library functions using -sDEFAULT_LIBRARY_FUNCS_TO_INCLUDE on the command line and export them via -sEXPORTED_RUNTIME_FUNCTIONS or -sEXPORTED_FUNCTIONS.
> > 
> > 
> > Are these options on the client side or something I need to change on the emscripten side?
> 
> Those options are what end uses would use if they want to call this function.

(Just like any other JS function)

sbc100 replied:
If you call a JS function from C then you don't need to do anything, it will automatically get included.

If you want to call a JS function on the `Module` object, or from and `EM_JS` block then you need to tell emscripten to include it somehow, otherwise it won't be included in the generated code.

Are you expecting end users to need to call `glfwSetHiDPIAware`, and if so from where?   For JS functions exposed to C/C++ they wouldn't need to do anything.  For JS functions exposed to the outside on the `Module.` object they would need to specify this function in a command line flag, just like any other JS library function.

ypujante replied:
I believe the ability to change dynamically either from C or Javascript is desirable:

* from C/C++, I envision it from a menu/checkbox (for example with ImGui I could provide both)
* from Javascript, the "shell" allow to request fullscreen for example, and so changing HiDPI in a similar manner could be an option.

If I have to keep one, I would pick C/C++

I am going to investigate your suggestions and see where that leads me.

sbc100 replied:
In general you don't need to pick one.. a JS library function can be called from both C/C++ and from external JS code.

The important thing is that if you want to call a function from outside the module you need to export it via a command line flag.   This is true for both native and JS function actually.  I general, if don't ask (via the command line) for a function to be exported it won't be.

ypujante replied:
For now I have pushed a change that only rename the function exported in Module to make it more specific to glfw.

```
  $GLFW__postset: `
    // making glfwSetHiDPIAware public for users of glfw (dynamic switching)
    Module["glfwSetHiDPIAware"] = GLFW.setHiDPIAware;`,
```

ypujante replied:
I am trying the export/import method and can't seem to make it work:

in `library_gflw.js` I added:

```javascript
  glfwSetHiDPIAware: (isHiDPIAware) => GLFW.setHiDPIAware(isHiDPIAware),
```

Now on the client side (C code), I added this:

```cpp
extern void glfwSetHiDPIAware(bool);
// ....
        glfwSetHiDPIAware(isHiDPIAware);
```

and the compilation flags:

```
-sDEFAULT_LIBRARY_FUNCS_TO_INCLUDE=glfwSetHiDPIAware -sEXPORTED_FUNCTIONS=glfwSetHiDPIAware
```

and getting this error:

```text
wasm-ld: error: CMakeFiles/imgui_raylib_cmake.dir/main.cpp.o: undefined symbol: glfwSetHiDPIAware(bool)
```

I guess I must be missing...

sbc100 replied:
You need to wrap the declaration in `extern "C"` which all JS library functions are C symbols, not C++.

Also you shouldn't need any extra command line flag to call it from C/C++.

ypujante replied:
See my most recent update about the changes

sbc100 commented on test/test_browser.py:
```diff
@@ -3018,8 +3018,12 @@ def test_glfw_events(self):
-  @no_wasm64('SDL2 + wasm64')
   @requires_graphics_hardware
+  def test_glfw3_hi_dpi_aware(self):
+    self.btest_exit('test_glfw3_hi_dpi_aware.c', args=['-sUSE_GLFW=3', '-sASYNCIFY', '-lglfw', '-lGL'])
```
Does this test use `-sASYNCIFY` ?  Also I don't think you need `-lglfw` if you use  `-sUSE_GLFW=3`.

ypujante replied:
Yes you are correct about `-sASYNCIFY`, that was a leftover of a previous development where I was calling sleep so that I could get a chance to look at the messages/console.

`-lglfw` is also not needed but I copy/pasted it from another glfw test.

I have removed both

ypujante commented:
Tests failing seems to have nothing with my changes:

```
FAIL [121.015s]: test_async_mainloop (test_browser.browser)
FAIL [120.957s]: test_async_jspi_wasm_bigint (test_browser.browser)
FAIL [121.059s]: test_async_returnvalue_em_js_bad (test_browser.browser)
FAIL [122.875s]: test_async_jspi (test_browser.browser)
FAIL [120.880s]: test_async_returnvalue_empty_list (test_browser.browser)
```

juj commented on src/library_glfw.js:
```diff
@@ -105,6 +107,7 @@ var LibraryGLFW = {
     versionString: null,
     initialTime: null,
     extensions: null,
+    devicePixelRatioMQS: null,
```
What is MQS short for?

ypujante replied:
It was supposed to be MediaQuerySearch but I realized that `window.matchMedia` returns a `MediaQueryList` not `MediaQuerySearch`... so I just renamed it `devicePixelRatioMQL` and added a comment for clarity

juj commented on src/library_glfw.js:
```diff
@@ -1227,9 +1394,20 @@ var LibraryGLFW = {
     Module["canvas"].addEventListener('dragover', GLFW.onDragover, true);
 
+    // Overriding implementation to account for HiDPI
+    Browser.requestFullscreen = GLFW.requestFullscreen
```
Missing `;` at end of line.

ypujante replied:
fixed

juj commented on src/library_glfw.js:
```diff
@@ -616,44 +619,46 @@ var LibraryGLFW = {
     },
 
-    onCanvasResize: (width, height) => {
+    onCanvasResize: (width, height, framebufferWidth, framebufferHeight) => {
```
Would be good to document here why there are two pairs of widthxheight and how they differ.

ypujante replied:
I just did.

juj commented on src/library_glfw.js:
```diff
@@ -1112,7 +1106,8 @@ var LibraryGLFW = {
       // Get non alive id
-      var win = new GLFW_Window(id, width, height, title, monitor, share);
+      const canvas = Module['canvas'];
+      var win = new GLFW_Window(id, canvas.clientWidth, canvas.clientHeight, canvas.width, canvas.height, title, monitor, share);
```
Looks like this changes the semantics of the existing `width` and `height` parameters passed to the `GLFW.createWindow(width, height)`  function, and the `GLFW_Window` ctor:

previously they were the WebGL front buffer pixel dimensions,
now after this PR `width` and `height` instead refer to the visible CSS size of the canvas.

Is that semantical change intentional?

juj replied:
Great, sounds good. LGTM.

ypujante commented:
The dreaded random tests failures are back:

```text
FAIL [122.412s]: test_async_2 (test_browser.browser64_4gb)
FAIL [120.975s]: test_async_asyncify (test_browser.browser64_4gb)
FAIL [121.017s]: test_async_bad_list (test_browser.browser64_4gb)
FAIL [120.942s]: test_async_compile (test_browser.browser64_4gb)
FAIL [121.248s]: test_async_in_pthread (test_browser.browser64_4gb)
```

But the primary test for this PR is passing:

```text
  test_glfw3_hi_dpi_aware (test_browser.browser64) ... Clearing existing test directory
cache:INFO: generating system asset: symbol_lists/2b7e561b0bade8e7f068e21c61f5fbc67208cca1.json... (this will be cached in "/root/cache/symbol_lists/2b7e561b0bade8e7f068e21c61f5fbc67208cca1.json" for subsequent builds)
cache:INFO:  - ok
ok (0.936s)
```

ypujante commented:
@juj as a side note regarding your comment:

>  b) how big is your appetite to further develop library_glfw.js? If you're enthusiastic for championing further development, maybe this might be the time to migrate library_glfw.js into its own repository and maybe use the ports system to distribute it? (that's ok if not, just bringing up the idea in general)

I have started looking into this... and toying around with implementation (c++ like your webgpu library)... :)

ypujante commented:
After merging with latest branch as of this morning, all tests are now passing. Yay!

sbc100 commented:
Great!  Landing this now.

ypujante commented:
@sbc100 Thank you! Let me know if you want me to update any release notes or documentation! (I can do a PR just for that purpose)

ypujante commented:
@juj I must admit I have been enjoying working on a glfw3 port... although I am far from finished, I already do support multiple canvases each with their own hi dpi aware flag! I had no idea if I could make it work, but it does (as can be seen in the screenshot). So exciting!!!

```cpp
// from main.cpp
  glfwWindowHint(GLFW_SCALE_TO_MONITOR, GLFW_TRUE);
  glfwWindowHintString(GLFW_EMSCRIPTEN_CANVAS_SELECTOR, "#canvas1");
  auto window1 = glfwCreateWindow(300, 200, "4k", nullptr, nullptr);

  glfwWindowHint(GLFW_SCALE_TO_MONITOR, GLFW_FALSE);
  glfwWindowHintString(GLFW_EMSCRIPTEN_CANVAS_SELECTOR, "#canvas2");
  auto window2 = glfwCreateWindow(300, 200, "2k", nullptr, nullptr);
```

```html
// from shell.html
<canvas class=emscripten id=canvas1 oncontextmenu=event.preventDefault() tabindex=-1></canvas>
<canvas class=emscripten id=canvas2 oncontextmenu=event.preventDefault() tabindex=-1></canvas>
```

<img width="681" alt="Screenshot 2023-12-09 at 11 47 07" src="https://github.com/emscripten-core/emscripten/assets/357116/fbd34530-6947-4500-82be-86a9822e7bbd">
